zebra-chain/src/primitives/redpallas/batch.rs

```diff
@@ -0,0 +1,265 @@
+    inner: Inner,
+}
+
+impl<'msg, M: AsRef<[u8]>>
+    From<(
+        VerificationKeyBytes<SpendAuth>,
+        Signature<SpendAuth>,
+        &'msg M,
+    )> for Item
+{
+    fn from(
+        (vk_bytes, sig, msg): (
+            VerificationKeyBytes<SpendAuth>,
+            Signature<SpendAuth>,
+            &'msg M,
+        ),
+    ) -> Self {
+        // Compute c now to avoid dependency on the msg lifetime.
+        let c = HStar::default()
+            .update(&sig.r_bytes[..])
+            .update(&vk_bytes.bytes[..])
+            .update(msg)
+            .finalize();
+        Self {
+            inner: Inner::SpendAuth { vk_bytes, sig, c },
+        }
+    }
+}
+
+impl<'msg, M: AsRef<[u8]>> From<(VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M)>
+    for Item
+{
+    fn from(
+        (vk_bytes, sig, msg): (VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M),
+    ) -> Self {
+        // Compute c now to avoid dependency on the msg lifetime.
+        let c = HStar::default()
+            .update(&sig.r_bytes[..])
+            .update(&vk_bytes.bytes[..])
+            .update(msg)
+            .finalize();
+        Self {
+            inner: Inner::Binding { vk_bytes, sig, c },
+        }
+    }
+}
+
+impl Item {
+    /// Perform non-batched verification of this `Item`.
+    ///
+    /// This is useful (in combination with `Item::clone`) for implementing fallback
+    /// logic when batch verification fails. In contrast to
+    /// [`VerificationKey::verify`](crate::VerificationKey::verify), which requires
+    /// borrowing the message data, the `Item` type is unlinked from the lifetime of
+    /// the message.
+    #[allow(non_snake_case)]
+    pub fn verify_single(self) -> Result<(), Error> {
+        match self.inner {
+            Inner::Binding { vk_bytes, sig, c } => VerificationKey::<Binding>::try_from(vk_bytes)
+                .and_then(|vk| vk.verify_prehashed(&sig, c)),
+            Inner::SpendAuth { vk_bytes, sig, c } => {
+                VerificationKey::<SpendAuth>::try_from(vk_bytes)
+                    .and_then(|vk| vk.verify_prehashed(&sig, c))
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+/// A batch verification context.
+pub struct Verifier {
+    /// Signature data queued for verification.
+    signatures: Vec<Item>,
+}
+
+impl Verifier {
+    /// Construct a new batch verifier.
+    pub fn new() -> Verifier {
+        Verifier::default()
+    }
+
+    /// Queue an Item for verification.
+    pub fn queue<I: Into<Item>>(&mut self, item: I) {
+        self.signatures.push(item.into());
+    }
+
+    /// Perform batch verification, returning `Ok(())` if all signatures were
+    /// valid and `Err` otherwise.
+    ///
+    /// The batch verification equation is:
+    ///
+    /// h_G * -[sum(z_i * s_i)]P_G + sum(\[z_i\]R_i + [z_i * c_i]VK_i) = 0_G
+    ///
+    /// which we split out into:
+    ///
+    /// h_G * -[sum(z_i * s_i)]P_G + sum(\[z_i\]R_i) + sum([z_i * c_i]VK_i) = 0_G
```

Comment:
You're missing an open paren after "h_G *" in both equations. I would suggest a correction but I don't know what the intended splitting is.

Reply:
I think we moved away from [h_g](<Point>) notation here so I'm going to leave this as-is